---
title: "Assignment 1 | 624 |"
author: "Kohl Kinning"
output: 
  html_notebook: 
    code_folding: none
    theme: cosmo
    toc: yes
---

## Part 1 --– SF-Seq read quality score distributions

Kohl    10_2G_both      17_3E_fox

The demultiplexed .fastq files are here: /projects/bgmp/2017_sequencing/demultiplexed/

1. *Using FastQC on Talapas, produce plots of quality score distributions for forward and reverse reads. Also, produce plots of the per-base N content, and comment on whether or not they are consistent with the quality score plots.*

2. *Run your quality score plotting script from the index hopping assignment. Describe how the FastQC quality score distribution plots compare to your own. If different, propose an explanation. Also, does the runtime differ? If so, why?*

```{r}
R1_10_2G = read.delim("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_1/q1.out", header = TRUE, sep = "\t")
R2_10_2G = read.delim("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_1/q2.out", header = TRUE, "\t")
R1_17_3E = read.delim("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_1/q3.out", header = TRUE, sep = "\t")
R2_17_3E = read.delim("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_1/q4.out", header = TRUE, sep = "\t")

plot(R1_10_2G$BPPosition, R1_10_2G$MeanScore)
plot(R2_10_2G$BPPosition, R2_10_2G$MeanScore)
plot(R1_17_3E$BPPosition, R1_17_3E$MeanScore)
plot(R2_17_3E$BPPosition, R2_17_3E$MeanScore)
```


## Part 2 – Adaptor trimming comparison

3. *Look into the adaptor trimming options for ```cutadapt```, ```process_shortreads```, and ```Trimmomatic``` (all on Talapas), and briefly describe the differences. Pick one of these to properly trim adapter sequences. Use default settings. What proportion of reads (both forward and reverse) was trimmed?* *Sanity check*: Use your Unix skills to search for the adapter sequences in your datasets and confirm the expected sequence orientations.
    + **cutadapt:** cutadapt removes adapter sequences from high-throughput sequencing reads. Replace "ADAPTER" with the actual sequence of your 3' adapter. IUPAC wildcard characters are supported. The reverse complement is *not* automatically searched. All reads from input.fastq will be written to output.fastq with the adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter sequences can be given (use further -a options), but only the best-matching adapter will be removed. 
    + **process_shortreads:** process_shortreads is part of the Stacks program. 
    + **Trimmomatic:** trimmomatic is a program written in java. A particular strength of Trimmomatic is that it retains the pairing of reads. It provides four output files when trimming paired end reads: trimmed paired and unpaired for both forward and reverse reads.
    
### Trimmomatic
#### Commands
**10_2G_both**
```{bash}
$ java -jar $EBROOTTRIMMOMATIC/trimmomatic-0.36.jar PE ../10_2G_both_S8_L008_R1_001.fastq ../10_2G_both_S8_L008_R2_001.fastq ../out_paired_10_2G_both_S8_L008_R1_001.fastq ../out_unpaired_10_2G_both_S8_L008_R1_001.fastq ../out_paired_10_2G_both_S8_L008_R2_001.fastq ../out_unpaired_10_2G_both_S8_L008_R2_001.fastq -phred33 ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
```
**17_3E_fox**
```{bash}
$ java -jar $EBROOTTRIMMOMATIC/trimmomatic-0.36.jar PE ../17_3E_fox_S13_L008_R1_001.fastq ../17_3E_fox_S13_L008_R2_001.fastq ../out_paired_17_3E_fox_S13_L008_R1_001.fastq ../out_unpaired_17_3E_fox_S13_L008_R1_001.fastq ../out_paired_17_3E_fox_S13_L008_R2_001.fastq ../out_unpaired_17_3E_fox_S13_L008_R2_001.fastq -phred33 ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
```

#### Results
**10_2G_both**
```{bash}
Using PrefixPair: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC' and 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT'
ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
Input Read Pairs: 81477069 Both Surviving: 75244573 (92.35%) Forward Only Surviving: 5957979 (7.31%) Reverse Only Surviving: 152833 (0.19%) Dropped: 121684 (0.15%)
```
**17_3E_fox**
```{bash}
Using PrefixPair: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC' and 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT'
ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
Input Read Pairs: 11784410 Both Surviving: 11034285 (93.63%) Forward Only Surviving: 708449 (6.01%) Reverse Only Surviving: 25910 (0.22%) Dropped: 15766 (0.13%)
TrimmomaticPE: Completed successfully
```


  
4. Plot the trimmed read length distributions for both forward and reverse reads (on the same plot). If necessary, consult Assignment 5 (Block 1) from Bi 623 to refresh your memory.

```{.bash}
$ awk 'NR % 4 == 2' out_paired_10_2G_both_S8_L008_R1_001.fastq | awk '{print length}' |sort | uniq -c > 10_2G_R1_lengths.txt
$ awk 'NR % 4 == 2' out_paired_10_2G_both_S8_L008_R2_001.fastq | awk '{print length}' |sort | uniq -c > 10_2G_R2_lengths.txt

$ awk 'NR % 4 == 2' out_paired_17_3E_fox_S13_L008_R1_001.fastq  | awk '{print length}' |sort | uniq -c > 17_3E_R1_lengths.txt
$ awk 'NR % 4 == 2' out_paired_17_3E_fox_S13_L008_R2_001.fastq  | awk '{print length}' |sort | uniq -c > 17_3E_R2_lengths.txt
```

```{r}
trimmed_len_R1_10_2G = read.table("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_2/10_2G_R1_lengths.txt")
trimmed_len_R2_10_2G = read.table("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_2/10_2G_R2_lengths.txt")

trimmed_len_R1_17_3E = read.table("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_2/17_3E_R1_lengths.txt")
trimmed_len_R1_17_3E = read.table("/Users/kohl/Documents/GradSchool/semester2/GenomicsLab/Assignments/sf-seq-qaa-kohlkopf/Part_2/17_3E_R1_lengths.txt")

colnames(trimmed_len_R1_10_2G)[1:2] = c("counts", "lengths")
colnames(trimmed_len_R2_10_2G)[1:2] = c("counts", "lengths")

plot(trimmed_len_R1_10_2G$counts~trimmed_len_R1_10_2G$lengths, 
     log = "y", col = "magenta4", 
     pch = 3, xlab = "read lengths", 
     ylab = "counts", 
     main = "Trimmed Data Distributions 10_2G Both ")
points(trimmed_len_R2_10_2G$counts~trimmed_len_R2_10_2G$lengths, col = "green4", pch = 5)

colnames(trimmed_len_R1_17_3E)[1:2] = c("counts", "lengths")
colnames(trimmed_len_R1_17_3E)[1:2] = c("counts", "lengths")

plot(trimmed_len_R1_17_3E$counts~trimmed_len_R1_17_3E$lengths, 
     log = "y", 
     col = "magenta4", pch = 3, xlab = "read lengths", 
     ylab = "counts", 
      main = "Trimmed Data Distributions 17_3E Fox")
points(trimmed_len_R1_17_3E$counts~trimmed_len_R1_17_3E$lengths, col = "green4", pch = 5)
```


5. Briefly describe whether the adaptor trimming results are consistent with the insert size distributions for your libraries. The size distribution information is in the Fragment Analyzer trace file on Github.

## Part 3 – rRNA reads and strand-specificity
6. Find publicly available mouse rRNA sequences and generate a gsnap database from them. Align the SF-Seq reads to your mouse rRNA database and report the proportion of reads that likely came from rRNAs.

7. Demonstrate convincingly that the SF-Seq data are from “strand-specific” RNA-Seq libraries. There are a number of possible strategies to address this problem, but you need only implement one. Report your evidence in numeric and graphical (e.g. a plot) forms.